File: qbopt/frontend/fppatches.py
from dataclasses import replace

from iced_x86 import Code

from qbopt.objectfile import omf
from qbopt.frontend.declen import decode
from qbopt.objectfile.module import Module


def native_records(module: Module, starts: frozenset[int]) -> list[omf.Record]:
    patches = sites(module, starts)
    removed: dict[int, list[tuple[int, int]]] = {}
    for fixup in omf.fixups(module.records):
        if fixup.seg == module.seg and fixup.offset in patches:
            removed.setdefault(id(fixup.record), []).append((fixup.lo, fixup.hi))
    records: list[omf.Record] = []
    for record in module.records:
        ranges = sorted(removed.get(id(record), ()))
        if not ranges:
            records.append(record)
            continue
        fragments = []
        cursor = 0
        for lo, hi in ranges:
            fragments.append(record.body[cursor:lo])
            cursor = hi
        fragments.append(record.body[cursor:])
        body = b"".join(fragments)
        if body:
            records.append(replace(record, body=body, raw=None))
    return records


def sites(module: Module, starts: frozenset[int]) -> frozenset[int]:
    names = omf.externals(module.records)
    result: set[int] = set()
    overrides = {"FIERQQ": 0x26, "FICRQQ": 0x2E, "FISRQQ": 0x36, "FIARQQ": 0x3E}
    for at, fixup in module.fixup_at.items():
        if (
            at not in starts
            or fixup.loc != omf.LOC_OFF16
            or fixup.selfrel
            or fixup.target != "external"
            or fixup.disp != 0
            or not 0 < fixup.index < len(names)
        ):
            continue
        name = names[fixup.index]
        if name == "FIWRQQ":
            if module.code[at : at + 2] == b"\x90\x9b" and at + 1 in starts:
                result.add(at)
            continue
        opcode = at + 1
        if name in overrides:
            if module.code[opcode : opcode + 1] != bytes([overrides[name]]):
                continue
            opcode += 1
        elif name != "FIDRQQ":
            continue
        if (
            module.code[at : at + 1] != b"\x9b"
            or at + 1 not in starts
            or opcode >= module.end
            or not 0xD8 <= module.code[opcode] <= 0xDF
        ):
            continue
        instruction = decode(module.code, at + 1)
        if instruction is not None and instruction.code != Code.INVALID:
            result.add(at)
    return frozenset(result)

File: src/frontend/fppatches.rs
//! Port of `qbopt/frontend/fppatches.py`: the FP emulator's own patch sites,
//! and the records without their fixups.

use std::collections::BTreeSet;
use std::rc::Rc;

use iced_x86::Code;

use crate::frontend::declen::decode;
use crate::objectfile::module::Module;
use crate::objectfile::omf::{self, Record};
use crate::support::hash::IndexMap;

pub fn native_records(module: &Module, starts: &BTreeSet<usize>) -> Vec<Rc<Record>> {
    let patches = sites(module, starts);
    // keyed by id(record)
    let mut removed: IndexMap<*const Record, Vec<(usize, usize)>> = IndexMap::default();
    for fixup in omf::fixups(&module.records) {
        if fixup.seg == Some(module.seg) && fixup.offset >= 0 && patches.contains(&(fixup.offset as usize)) {
            removed.entry(Rc::as_ptr(&fixup.record)).or_default().push((fixup.lo, fixup.hi));
        }
    }
    let mut records: Vec<Rc<Record>> = Vec::new();
    for record in &module.records {
        let mut ranges = removed.get(&Rc::as_ptr(record)).cloned().unwrap_or_default();
        ranges.sort();
        if ranges.is_empty() {
            records.push(record.clone());
            continue;
        }
        let mut fragments: Vec<&[u8]> = Vec::new();
        let mut cursor = 0;
        for (lo, hi) in ranges {
            fragments.push(slice(&record.body, cursor, lo));
            cursor = hi;
        }
        fragments.push(slice(&record.body, cursor, record.body.len()));
        let body = fragments.concat();
        if !body.is_empty() {
            records.push(Rc::new(Record { r#type: record.r#type, body, raw: None }));
        }
    }
    records
}

pub fn sites(module: &Module, starts: &BTreeSet<usize>) -> BTreeSet<usize> {
    let names = omf::externals(&module.records);
    let mut result: BTreeSet<usize> = BTreeSet::new();
    let overrides = |name: &str| match name {
        "FIERQQ" => Some(0x26u8),
        "FICRQQ" => Some(0x2E),
        "FISRQQ" => Some(0x36),
        "FIARQQ" => Some(0x3E),
        _ => None,
    };
    for (&at, fixup) in &module.fixup_at {
        let at = at as usize;
        if !starts.contains(&at)
            || fixup.loc != omf::LOC_OFF16
            || fixup.selfrel
            || fixup.target != "external"
            || fixup.disp != 0
            || !(0 < fixup.index && (fixup.index as usize) < names.len())
        {
            continue;
        }
        let name = names[fixup.index as usize].as_str();
        if name == "FIWRQQ" {
            if slice(&module.code, at, at + 2) == b"\x90\x9b" && starts.contains(&(at + 1)) {
                result.insert(at);
            }
            continue;
        }
        let mut opcode = at + 1;
        if let Some(prefix) = overrides(name) {
            if slice(&module.code, opcode, opcode + 1) != [prefix] {
                continue;
            }
            opcode += 1;
        } else if name != "FIDRQQ" {
            continue;
        }
        if slice(&module.code, at, at + 1) != b"\x9b"
            || !starts.contains(&(at + 1))
            || opcode as i64 >= module.end
            || !(0xD8..=0xDF).contains(&module.code[opcode])
        {
            continue;
        }
        let instruction = decode(&module.code, at + 1);
        if instruction.is_some_and(|instruction| instruction.code() != Code::INVALID) {
            result.insert(at);
        }
    }
    result
}

/// `b[lo:hi]` for non-negative bounds: clamped, and empty where `hi < lo`.
fn slice(b: &[u8], lo: usize, hi: usize) -> &[u8] {
    let hi = hi.min(b.len());
    if lo >= hi { &[] } else { &b[lo..hi] }
}
